chrome/content/zotero/feedSettings.js

```diff
@@ -0,0 +1,188 @@
+    along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+    
+    ***** END LICENSE BLOCK *****
+*/
+
+//////////////////////////////////////////////////////////////////////////////
+//
+// Zotero_Feed_Settings
+//
+//////////////////////////////////////////////////////////////////////////////
+
+var Zotero_Feed_Settings = new function() {
+	let urlIsValid = true,
+		data = null,
+		feedReader = null,
+		urlTainted = false;
+	
+	let cleanURL = function(url) {
+		url = url.trim();
+		if (!url) return;
+		
+		let ios = Components.classes["@mozilla.org/network/io-service;1"]
+			.getService(Components.interfaces.nsIIOService);
+		
+		let cleanUrl;
+		try {
+			let uri = ios.newURI(url, null, null);
+			if (uri.scheme != 'http' && uri.scheme != 'https') {
+				Zotero.debug(uri.scheme + " is not a supported protocol for feeds.");
+			}
+			
+			cleanUrl = uri.spec;
+		} catch (e) {
+			if (e.result == Components.results.NS_ERROR_MALFORMED_URI) {
+				// Assume it's a URL missing "http://" part
+				try {
+					cleanUrl = ios.newURI('http://' + url, null, null).spec;
+				} catch (e) {}
+			}
+			throw e;
+		}
+		
+		if (!cleanUrl) return;
+		
+		if (/^https?:\/\/[^\/\s]+\/\S/.test(cleanUrl)) return cleanUrl;
+	};
+	
+	this.init = function() {
+		this.toggleAdvancedOptions(false);
+		
+		data = window.arguments[0];
+		
+		if (data.url) {
+			document.getElementById('feed-url').value = data.url;
+		}
+		
+		if (!data.url) {
+			this.invalidateUrl();
+		} else {
+			// Do not allow to change URL for existing feed
+			document.getElementById('feed-url').readOnly = true;
+		}
+		
+		if (data.title) {
+			document.getElementById('feed-title').value = data.title;
+		}
+		
+		let ttl;
+		if (data.ttl !== undefined) {
+			ttl = Math.floor(data.ttl / 60);
+		} else {
+			ttl = 1;
+		}
+		document.getElementById('feed-ttl').value = ttl;
+		
+		let cleanAfter = data.cleanAfter;
+		if (cleanAfter === undefined) cleanAfter = 2;
+		document.getElementById('feed-cleanAfter').value = cleanAfter;
+		
+		if (data.url && !data.urlIsValid) {
+			this.validateUrl();
+		}
+	};
+	
+	this.invalidateUrl = function() {
+		urlTainted = true;
+		if (feedReader) {
+			feedReader.terminate();
+			feedReader = null;
+		}
+		
+		if (!urlIsValid) return;
+		
+		urlIsValid = false;
+		document.getElementById('feed-title').disabled = true;
+		document.getElementById('feed-ttl').disabled = true;
+		document.getElementById('feed-cleanAfter').disabled = true;
+		document.documentElement.getButton('accept').disabled = true;
+	};
+	
+	this.validateUrl = Zotero.Promise.coroutine(function* () {
+		if (feedReader) {
+			feedReader.terminate();
+			feedReader = null;
+		}
+		
+		let url = cleanURL(document.getElementById('feed-url').value);
+		urlTainted = false;
+		if (!url) return;
+		
+		try {
+			let fr = feedReader = new Zotero.FeedReader(url);
+			yield fr.process();
+			let feed = fr.feedProperties;
+			if (feedReader !== fr || urlTainted) return;
```

Comment:
Maybe make a comment about the logic here, since it's not very obvious. The idea is that if `validateUrl`/`invalidateUrl` were called some time during `fr.process()`, then we should abort further processing, since this is an outdated "session"